espresso/environment/4_confirmation_integrity_with_reorgs_test.go

```diff
@@ -0,0 +1,217 @@
+	require.NoError(t, err)
+	unsafeL2BlockNumber := status.SafeL2.Number + 1
+
+	nonce := uint64(0)
+	addressAlice := system.Cfg.Secrets.Addresses().Alice
+
+	require.NoError(t, err)
+	for (l1Height - l1HeightStart) < system.Cfg.L1FinalizedDistance {
+		height := uint64(i) + unsafeL2BlockNumber
+
+		//Send some transactions to fill the batches
+		receipt := helpers.SendL2TxWithID(t, system.Cfg.L2ChainIDBig(), l2Seq, system.Cfg.Secrets.Bob, func(opts *helpers.TxOpts) {
+			opts.Nonce = nonce
+			opts.ToAddr = &addressAlice
+			opts.Value = new(big.Int).SetUint64(1)
+		})
+		nonce++
+		log.Info("Receipt", "value", receipt)
+
+		l2Head, err := geth.WaitForBlockToBeSafe(new(big.Int).SetUint64(height), l2Verif, 10*time.Second)
+		require.NoError(t, err)
+		
+		if err != nil {
+			continue
+		} else { // Insert new batch in the list
+
+			batch, l2HeadL1Info, err := derive.BlockToSingularBatch(system.RollupCfg(), l2Head)
+			require.NoError(t, err)
+			log.Info("l2HeadL1Info", "value", l2HeadL1Info)
+
+			batchHash := BatchHash(batch)
+
+			t.Log("New element inserted", "value", batchHash, "list length", len(batches))
+			batches = append(batches, batchHash)
+
+			i++
+		}
+
+		l1Height, err = l1Client.BlockNumber(ctx)
+		require.NoError(t, err)
+	}
+
+	return batches, l1HeightStart, unsafeL2BlockNumber
+}
+
+// This collect the first N L2 blocks from a specific height. Collected blocks are guaranteed to be safe
+// @param ctx,t,system standard parameters to execute a test.
+// @param l2Verif OP node to fetch the safe blocks.
+// @apram n number of blocks to fetch.
+// @param startIndex initial height of the L2 chain to start fetching blocks from.
+func collectFirstNL2SafeBlocks(ctx context.Context, t *testing.T, system *e2esys.System, l2Verif *ethclient.Client, n int, startIndex uint64) []string {
+	var batches []string
+
+	for i := 0; i < n; i++ {
+		height := startIndex + uint64(i)
+		_, err := geth.WaitForBlockToBeSafe(big.NewInt(int64(height)), l2Verif, 2*time.Minute)
+		require.NoError(t, err)
+
+		l2Head, err := l2Verif.BlockByNumber(ctx, new(big.Int).SetUint64(height))
+		require.NoError(t, err)
+
+		batch, _, err := derive.BlockToSingularBatch(system.RollupCfg(), l2Head)
+		batchHash := BatchHash(batch)
+		batches = append(batches, batchHash)
+
+	}
+	return batches
+}
+
+// Main logic of the test:
+// 1. Collect some unsafe batches in list L.
+// 2. Do the reorg.
+// 3. Collect the batches into list L'.
+// 4. Check that L=L'.
+func run(ctx context.Context, t *testing.T, system *e2esys.System) {
+	l2Seq := system.NodeClient(e2esys.RoleSeq)
+	l2Verif := system.NodeClient(e2esys.RoleVerif)
+	l1Client := system.NodeClient(e2esys.RoleL1)
+
+	var unsafeL2Height uint64
+	var l1Height uint64
+
+	// Wait for batcher to start advancing L2 head
+	_, err := geth.WaitForBlockToBeSafe(big.NewInt(2), l2Seq, 2*time.Minute)
+	if have, want := err, error(nil); have != want {
+		t.Fatalf("L2 isn't progressing:\nhave:\n\t%v\nwant:\n\t%v", have, want)
+	}
```

Comment:
```suggestion
	require.NoError(t, err, "L2 isn't progressing as expected")
```

We are using this syntax in many places though, so it's okay to do a cleanup in a separate PR.